File: boj/1533.cpp
//
//  main.cpp
//  boj1533
//
//  Created by 최연웅 on 2018. 9. 22..
//  Copyright © 2018년 최연웅. All rights reserved.
//

#include <iostream>

#define DIV 1000003

using namespace std;

int** alloc_matrix(int n)
{
    int **m = (int**)malloc(n*sizeof(int*));
    for(int i = 0; i < n; i++) {
        m[i] = (int*)malloc(n*sizeof(int));
    }
    return m;
}

void free_matrix(int **m, int n)
{
    for(int i = 0; i < n; i++) {
        free(m[i]);
    }
    free(m);
}

void copy(int **dst, int **src, int n)
{
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < n; j++) {
            dst[i][j] = src[i][j];
        }
    }
}

int** matmul(int **a, int **b, int n)
{
    int **c = alloc_matrix(n);
    
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < n; j++) {
            int sum = 0;
            for(int k = 0; k < n; k++) {
                sum += (a[i][k] % DIV) * (b[k][j] % DIV) % DIV;
            }
            c[i][j] = sum;
        }
    }
    
    return c;
}

int main(int argc, const char * argv[]) {
    // insert code here...
    int **d;
    int g[10][10];
    int n, s, e, t;
    cin >> n >> s >> e >> t;
    s--; e--;
    for(int i = 0; i < n; i++) {
        int tmp;
        cin >> tmp;
        for(int j = n-1; j >= 0; j--) {
            g[i][j] = tmp % 10;
            tmp /= 10;
        }
    }
    
    int cnt = n;
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < n; j++) {
            if(g[i][j] > 1)
                cnt += g[i][j] - 1;
        }
    }
    
    d = alloc_matrix(cnt);
    
    int next = n;
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < n; j++) {
            if(g[i][j] > 1) {
                int former = i;
                for(int k = 1; k < g[i][j]; k++) {
                    d[former][next] = 1;
                    former = next++;
                }
                d[former][j] = 1;
            }
            else {
                d[i][j] = g[i][j];
            }
        }
    }
    
    int **m[30], b[30], digit = 0, t0 = t;
    
    while(t0 > 0) {
        b[digit] = t0 % 2;
        t0 /= 2;
        m[digit] = alloc_matrix(cnt);
        copy(m[digit], d, cnt);
        int **tmp = matmul(d, d, n);
        free_matrix(d, n);
        d = tmp;
        
        digit++;
    }
    
    for(int k = 0; k < digit; k++) {
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                printf("%d ", m[k][i][j]);
            }
            printf("\n");
        }
        printf("\n");
    }
    
    free_matrix(d, cnt);
    for(int i = 0; i < digit; i++) {
        free_matrix(m[i], cnt);
    }
    
    return 0;
}

File: boj/1533.py
import math

DIV: int = 1000003


def matmul(a, b):
    aw = a[0].__len__()
    ah = a.__len__()
    bw = b[0].__len__()
    c = [[0] * bw for i in range(ah)]
    for i in range(ah):
        for j in range(bw):
            c[i][j] = sum([(a[i][k] % DIV) * (b[k][j] % DIV) % DIV for k in range(aw)])

    return c


n, s, e, t = [int(num) for num in input().split()]
s -= 1
e -= 1
graph = [[0] * n for i in range(n)]

for i in range(n):
    line = input()
    for j in range(n):
        graph[i][j] = int(line[j])

cnt = n
for i in range(n):
    for j in range(n):
        if graph[i][j] > 1:
            cnt += graph[i][j] - 1

d = [[0] * cnt for i in range(cnt)]

new = n
for i in range(n):
    for j in range(n):
        if graph[i][j] > 1:
            former = i
            for k in range(1, graph[i][j]):
                d[former][new] = 1
                former = new
                new += 1
            d[former][j] = 1
        else:
            d[i][j] = graph[i][j]

# t를 이진수로 바꾸는 작업
# b[0] : 2^0 자리, ...
m = []
b = []
t0: int = t
while t0 > 0:
    b.append(t0 % 2)
    t0 = int(t0 / 2)
    m.append(d)
    d = matmul(d, d)

identity = [[0] * cnt for i in range(cnt)]
for i in range(cnt):
    identity[i][i] = 1

ret = identity

for l in range(m.__len__()):
    if b[l]:
        ret = matmul(ret, m[l])

print(ret[s][e] % DIV)

'''9 1 3 100000000
012341232
102312312
230431021
123042314
402301234
123430132
012304021
203234005
012340020

answer = 872914
'''